src/Microsoft.Diagnostics.NETCore.Client/DiagnosticsServerRouter/DiagnosticsServerRouterFactory.cs

```diff
@@ -255,33 +253,148 @@ public async Task<Stream> AcceptTcpStreamAsync(CancellationToken token)
+        }
+
+        public override void Reset()
+        {
+            if (Endpoint != null)
+            {
+                _tcpServer.RemoveConnection(NetServerEndpointInfo.RuntimeInstanceCookie);
+                ResetEnpointInfo();
+            }
+        }
+
+        protected override Task<IpcEndpointInfo> AcceptAsyncImpl(CancellationToken token) => _tcpServer.AcceptAsync(token);
+        public override string ServerAddress => _tcpServerAddress;
+        public override string ServerTransportName => "TCP";
+
+
+        public override void CreatedNewServer(EndPoint localEP)
         {
             if (localEP is IPEndPoint ipEP)
                 _tcpServerAddress = _tcpServerAddress.Replace(":0", string.Format(":{0}", ipEP.Port));
         }
     }
 
+    /// <summary>
+    /// This class represent a WebSocket server endpoint used when building up router instances.
+    /// </summary>
+    internal class WebSocketServerRouterFactory : NetServerRouterFactory
+    {
+
+        private readonly string _webSocketURL;
+
+        ReversedDiagnosticsServer _webSocketServer;
+
+        public string WebSocketURL => _webSocketURL;
+
+        public static WebSocketServerRouterFactory CreateDefaultInstance(string webSocketURL, int runtimeTimeoutMs, ILogger logger)
+        {
+            return new WebSocketServerRouterFactory(webSocketURL, runtimeTimeoutMs, logger);
+        }
+
+        public WebSocketServerRouterFactory(string webSocketURL, int runtimeTimeoutMs, ILogger logger) : base(runtimeTimeoutMs, logger)
+        {
+
+            _webSocketURL = string.IsNullOrEmpty(webSocketURL) ? "ws://127.0.0.1:8088/diagnostics" : webSocketURL;
+
+            _webSocketServer = new ReversedDiagnosticsServer(_webSocketURL, ReversedDiagnosticsServer.Kind.WebSocket);
+            _webSocketServer.TransportCallback = this;
+        }
+
+        public override void Start()
+        {
+            Logger.LogInformation("Starting web socket server");
+            _webSocketServer.Start();
+        }
+
+        public override async Task Stop()
+        {
+            Logger.LogInformation("Stopping web socket server");
+            await _webSocketServer.DisposeAsync().ConfigureAwait(false);
+        }
+
+        public override void Reset()
+        {
+            if (Endpoint != null)
+            {
+                Logger.LogInformation("Resetting the web socket server");
+                _webSocketServer.RemoveConnection(NetServerEndpointInfo.RuntimeInstanceCookie);
+                ResetEnpointInfo();
+            }
+        }
+
+        protected override Task<IpcEndpointInfo> AcceptAsyncImpl(CancellationToken token) => _webSocketServer.AcceptAsync(token);
+        public override string ServerAddress => WebSocketURL;
+        public override string ServerTransportName => "WebSocket";
+
+        public override void CreatedNewServer(EndPoint localEP)
+        {
+            // anything to do here?
```

Comment:
You only need to do something here in case you would like to get endpoint info IPCServerTransport after it has bound and started to listen on endpoint. In TcpServerRouterFactory it is used if dynamic ports are used, so it will then update the server address with the actually assigned port number so it can be logged.